packages/flutter_test/lib/src/test_text_input.dart

```diff
@@ -64,8 +63,20 @@ class TestTextInput {
   /// Whether this [TestTextInput] is registered with [SystemChannels.textInput].
   ///
   /// Use [register] and [unregister] methods to control this value.
-  bool get isRegistered => _isRegistered;
-  bool _isRegistered = false;
+  // Someone could have unregistered us by calling
+  // `SystemChannels.textInput.setMockMethodCallHandler` in the test. Actually
+  // check if we're registered by calling the method, which should work
+  // immediately if we are registered since that method does no real async work.
+  bool get isRegistered {
+    final int oldLogLength = log.length;
+    unawaited(SystemChannels.textInput.invokeMethod<void>('__isRegistered'));
+    if (log.length == oldLogLength) {
+      return false;
+    }
+    log.removeLast();
+    return true;
+  }
```

Comment:
I can move this to a separate PR if that works better - it's not needed for this issue.